src/Nethermind/Nethermind.Trie/BatchedTrieVisitor.cs

```diff
@@ -0,0 +1,346 @@
+    private ArrayPool<Job> _jobArrayPool = ArrayPool<Job>.Create();
+    private ArrayPool<(TrieNode, SmallTrieVisitContext)> _trieNodePool = ArrayPool<(TrieNode, SmallTrieVisitContext)>.Create();
+
+    private readonly int _maxBatchSize;
+    private readonly long _partitionCount;
+    private readonly CompactStack<Job>[] _partitions;
+    private readonly long _targetCurrentItems;
+
+    private long _activeJobs;
+    private long _queuedJobs;
+    private bool _failed;
+    private long _currentPointer;
+
+    private readonly ITrieNodeResolver _resolver;
+    private readonly ITreeVisitor _visitor;
+
+    public BatchedTrieVisitor(
+        ITreeVisitor visitor,
+        ITrieNodeResolver resolver,
+        VisitingOptions visitingOptions)
+    {
+        _visitor = visitor;
+        _resolver = resolver;
+
+        _maxBatchSize = 128;
+
+        // The keccak + context itself should be 40 byte. But the measured byte seems to be 52 from GC stats POV.
+        // The * 2 is just margin. RSS is still higher though, but that could be due to more deserialization.
+        long recordSize = 52 * 2;
+        long recordCount = visitingOptions.FullScanMemoryBudget / recordSize;
+
+        // Generally, at first, we want to attempt to maximize number of partition. This tend to increase throughput
+        // compared to increasing batch size.
+        _partitionCount = recordCount / _maxBatchSize;
+
+        // 3000 is about the num of file for state on mainnet, so we assume 4000 for an unpruned db. Multiplied by
+        // a reasonable num of thread we want to confine to a file. If its too high, the overhead of looping through the
+        // stack can get a bit high at the end of the visit. But then again its probably not much.
+        long maxPartitionCount = 4000 * Math.Min(4, visitingOptions.MaxDegreeOfParallelism);
+
+        if (_partitionCount > maxPartitionCount)
+        {
+            _partitionCount = maxPartitionCount;
+            _maxBatchSize = (int)(recordCount / _partitionCount);
+        }
+
+        // Calculating estimated pool margin at about 5% of total working size. The working set size fluctuate a bit so
+        // this is to reduce allocation.
+        long estimatedPoolMargin = (long)(((double)recordCount / 128) * 0.05);
+        ObjectPool<CompactStack<Job>.Node> jobPool = new DefaultObjectPool<CompactStack<Job>.Node>(
+            new CompactStack<Job>.ObjectPoolPolicy(128), (int)estimatedPoolMargin);
+
+        _currentPointer = 0;
+        _queuedJobs = 0;
+        _activeJobs = 0;
+        _targetCurrentItems = _partitionCount * _maxBatchSize;
+
+        // This need to be very small
+        _partitions = new CompactStack<Job>[_partitionCount];
+        for (int i = 0; i < _partitionCount; i++)
+        {
+            _partitions[i] = new CompactStack<Job>(jobPool);
+        }
+    }
+
+    // Determine the locality of the key. I guess if you use paprika or something, you'd need to modify this.
```

Comment:
@Scooletz :)

Reply:
I don't think (and I could be wrong here), it would benefit paprika. As the nodes that is written at around the same time wil naturally have similar offset with paprika.